### PR TITLE
[CI Visibility] Manual Api and Go/Testing integration

### PR DESCRIPTION
**Pull request by tonyredondo**
### What does this PR do?

This PR is a continuation of #2739 and adds the first integrations, containing:
- A manual api to create `TestSessions`, `TestModules`, `TestSuites` and `Tests`
- A first integration for instrumenting tests from the `go/testing` package. (using the manual api)

All current API are private inside the internal folder. We are planning to use them later by using the [orchestrion project](https://github.com/DataDog/orchestrion) to do automatic instrumentation.

### Motivation

This is the final step to make everything work. After this PR we can instrument tests with CI Visibility product.

### Stacked PRs

1. #2736 
2. #2739
3. #2742 | **[This one]**

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: internal/civisibility/integrations/manual_api_ddtestsuite.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+// Test Suite
+
+// Ensures that tslvTestSuite implements the DdTestSuite interface.
+var _ DdTestSuite = (*tslvTestSuite)(nil)
+
+// tslvTestSuite implements the DdTestSuite interface and represents a suite of tests within a module.
+type tslvTestSuite struct {
+	ciVisibilityCommon
+	module  *tslvTestModule
+	suiteId uint64
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_ddtestsuite.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/manual_api_ddtestsuite.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	// Suite tags should include module and session tags so the backend can calculate the module and session fingerprint from the suite.
+	suiteTags := append(module.tags, tracer.Tag(constants.TestSuite, name))
+	testOpts := append(fillCommonTags([]tracer.StartSpanOption{
+		tracer.ResourceName(resourceName),
+		tracer.SpanType(constants.SpanTypeTestSuite),
+		tracer.StartTime(startTime),
+	}), suiteTags...)
+
+	span, ctx := tracer.StartSpanFromContext(context.Background(), operationName, testOpts...)
+	suiteId := span.Context().SpanID()
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_ddtestsuite.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/manual_api_mocktracer_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+func createDDTest(now time.Time) (DdTestSession, DdTestModule, DdTestSuite, DdTest) {
+	session, module, suite := createDDTestSuite(now)
+	test := suite.CreateTestWithStartTime("my-test", now)
+	test.SetTag("my-tag", "my-value")
+	return session, module, suite, test
+}
+
+func commonAssertions(assert *assert.Assertions, now time.Time, sessionSpan mocktracer.Span) {
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_mocktracer_test.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/civisibility.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		// Preload the CodeOwner file
+		_ = utils.GetCodeOwners()
+
+		// Preload all CI, Git, and CodeOwners tags.
+		ciTags := utils.GetCiTags()
+
+		// Check if DD_SERVICE has been set; otherwise default to the repo name (from the spec).
+		var opts []tracer.StartOption
+		if v := os.Getenv("DD_SERVICE"); v == "" {
+			if repoUrl, ok := ciTags[constants.GitRepositoryURL]; ok {
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/civisibility.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/manual_api_ddtestmodule.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+// Test Module
+
+// Ensures that tslvTestModule implements the DdTestModule interface.
+var _ DdTestModule = (*tslvTestModule)(nil)
+
+// tslvTestModule implements the DdTestModule interface and represents a module within a test session.
+type tslvTestModule struct {
+	ciVisibilityCommon
+	session   *tslvTestSession
+	moduleId  uint64
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_ddtestmodule.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/manual_api_ddtestmodule.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	}...)
+
+	testOpts := append(fillCommonTags([]tracer.StartSpanOption{
+		tracer.ResourceName(resourceName),
+		tracer.SpanType(constants.SpanTypeTestModule),
+		tracer.StartTime(startTime),
+	}), moduleTags...)
+
+	span, ctx := tracer.StartSpanFromContext(context.Background(), operationName, testOpts...)
+	moduleId := span.Context().SpanID()
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_ddtestmodule.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/manual_api_ddtestsession.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+// Test Session
+
+// Ensures that tslvTestSession implements the DdTestSession interface.
+var _ DdTestSession = (*tslvTestSession)(nil)
+
+// tslvTestSession implements the DdTestSession interface and represents a session for a set of tests.
+type tslvTestSession struct {
+	ciVisibilityCommon
+	sessionId        uint64
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_ddtestsession.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/manual_api_ddtestsession.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	testOpts := append(fillCommonTags([]tracer.StartSpanOption{
+		tracer.ResourceName(resourceName),
+		tracer.SpanType(constants.SpanTypeTestSession),
+		tracer.StartTime(startTime),
+	}), sessionTags...)
+
+	span, ctx := tracer.StartSpanFromContext(context.Background(), operationName, testOpts...)
+	sessionId := span.Context().SpanID()
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/manual_api_ddtestsession.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing_test.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+	// (*M)(m).Run() cast m to gotesting.M and just run
+	// or use a helper method gotesting.RunM(m)
+
+	// os.Exit((*M)(m).Run())
+	os.Exit(RunM(m))
+}
+
+// TestMyTest02 demonstrates instrumentation of InternalTests
+func TestMyTest01(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing_test.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/coverage_post119.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	"strconv"
+	"testing"
+	_ "unsafe"
+)
+
+// runtime_coverage_processCoverTestDirInternal is an internal runtime function used to process coverage data.
+// This declaration uses go:linkname to access the unexported function from the runtime package.
+//
+//go:linkname runtime_coverage_processCoverTestDirInternal runtime/coverage.processCoverTestDirInternal
+func runtime_coverage_processCoverTestDirInternal(dir string, cfile string, cm string, cpkg string, w io.Writer) error
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/coverage_post119.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing.go [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+				commonInfo: commonInfo{
+					moduleName: moduleName,
+					suiteName:  suiteName,
+					testName:   test.Name,
+				},
+			}
+
+			// Initialize module and suite counters if not already present.
+			if _, ok := modulesCounters[moduleName]; !ok {
+				var v int32 = 0
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing.go [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+				modulesCounters[moduleName] = &v
+			}
+			// Increment the test count in the module.
+			atomic.AddInt32(modulesCounters[moduleName], 1)
+
+			if _, ok := suitesCounters[suiteName]; !ok {
+				var v int32 = 0
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+				commonInfo: commonInfo{
+					moduleName: moduleName,
+					suiteName:  suiteName,
+					testName:   benchmark.Name,
+				},
+			}
+
+			// Initialize module and suite counters if not already present.
+			if _, ok := modulesCounters[moduleName]; !ok {
+				var v int32 = 0
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing.go [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"time"
+	_ "unsafe"
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing_test.go [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+		// Create a new HTTP request
+		req, err := http.NewRequest("GET", s.URL+"/hello/world", nil)
+		if err != nil {
+			t.FailNow()
+		}
+
+		// Use the span context here so the http span will appear as a child of the test
+		req = req.WithContext(ctx)
+
+		_, _ = client.Do(req)
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing_test.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/gotesting/testing_test.go [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+		req, err := http.NewRequest("GET", s.URL+"/hello/world", nil)
+		if err != nil {
+			t.FailNow()
+		}
+
+		// Use the span context here so the http span will appear as a child of the test
+		req = req.WithContext(ctx)
+
+		_, _ = client.Do(req)
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing_test.go]: fixed in [a6dc968](https://github.com/DataDog/dd-trace-go/pull/2742/commits/a6dc9686014d9aee473563dce03b1a0bb0a75734)

File: internal/civisibility/integrations/civisibility.go [hunk 17]
```diff
@@ -0,0 +1,118 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/civisibility.go:1]: it was a rebase error, fixed.

File: internal/civisibility/integrations/civisibility.go [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+		// Initialize the mocktracer
+		mTracer = mocktracer.Start()
+	})
+	return mTracer
+}
+
+func internalCiVisibilityInitialization(tracerInitializer func([]tracer.StartOption)) {
+	ciVisibilityInitializationOnce.Do(func() {
+		// Since calling this method indicates we are in CI Visibility mode, set the environment variable.
+		_ = os.Setenv(constants.CiVisibilityEnabledEnvironmnetVariable, "1")
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/civisibility.go]: it was a rebase error, fixed.

File: internal/civisibility/integrations/gotesting/testing.go [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+	"reflect"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/constants"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations"
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/testing.go:20]: it was a rebase error, fixed.

File: internal/civisibility/integrations/civisibility.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (tonyredondo) [reply internal/civisibility/integrations/civisibility.go]: it was a rebase error, fixed.

File: internal/civisibility/integrations/gotesting/reflections_test.go [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+	// Dereference the pointer to get the actual value
+	actualValue := (*string)(ptr)
+	if *actualValue != mockStruct.privateField {
+		t.Fatalf("Expected 'testValue', got %s", *actualValue)
+	}
+
+	// Modify the value through the pointer
+	*actualValue = "modified value"
+	if *actualValue != mockStruct.privateField {
+		t.Fatalf("Expected 'testValue', got %s", *actualValue)
```
Comment (ManuelPalenzuelaDD) [inline internal/civisibility/integrations/gotesting/reflections_test.go:43]: Should this be `Expected 'modified value', got %s`?

**Review by ManuelPalenzuelaDD — APPROVED** [review]
LGTM, left a comment

File: internal/civisibility/integrations/gotesting/reflections_test.go (hunk 20 above)
Comment (tonyredondo) [reply internal/civisibility/integrations/gotesting/reflections_test.go]: you're totally right, I fixed that in [4b3cda6](https://github.com/DataDog/dd-trace-go/pull/2742/commits/4b3cda60a0994af37829bea5fa627dc0ac613566)

**Review by darccio — CHANGES_REQUESTED** [review]

**Review by darccio — APPROVED** [review]

File: internal/civisibility/integrations/civisibility.go [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+func InitializeCiVisibilityMock() mocktracer.Tracer {
+	internalCiVisibilityInitialization(func([]tracer.StartOption) {
+		// Initialize the mocktracer
+		mTracer = mocktracer.Start()
+	})
+	return mTracer
+}
+
+func internalCiVisibilityInitialization(tracerInitializer func([]tracer.StartOption)) {
+	ciVisibilityInitializationOnce.Do(func() {
```
[the remark this answers is not in the document]
Comment (liashenko) [reply internal/civisibility/integrations/civisibility.go:57]: I would keep it as it is if you don't mind 😃

**Review by darccio — APPROVED** [review]

**Pull request merged**